// stitch-cli/lib/commands/refresh/remote-schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const client_1 = require("../../client");
const get_env_info_1 = require("../../utils/get-env-info");
class ApplyResources extends command_1.Command {
    constructor() {
        super(...arguments);
        this.verboseMode = false;
    }
    async run() {
        const { args, flags } = this.parse(ApplyResources);
        this.verboseMode = flags['verbose'];
        try {
            this.log(`Refreshing remote schema...`);
            const { result: { success }, } = await client_1.refreshRemoteSchema(args.remoteServerUrl, {
                registryUrl: flags['registry-url'],
                authorizationHeader: flags['authorization-header'],
            }, {
                timeout: flags.timeout,
            });
            if (success) {
                this.log(`Remote schema from ${args.remoteServerUrl} was refreshed successfully.`);
            }
            else {
                throw new Error('Something went wrong');
            }
        }
        catch (e) {
            this.error(`Refresh remote schema failed. ${e}

          ${get_env_info_1.default(this.config, 'refresh:remote-schema', flags['registry-url'], flags['authorization-header'])}`, { ...e, exit: true });
        }
    }
    async trace(message, ...args) {
        if (this.verboseMode) {
            this.log(message, ...args);
        }
    }
}
exports.default = ApplyResources;
ApplyResources.description = 'Refresh remote schema';
ApplyResources.examples = [
    `$ stitch refresh:remote-schema http://remote-graphql-server/graphql
Remote schema refreshed successfully!
`,
];
ApplyResources.flags = {
    'registry-url': command_1.flags.string({ required: true, env: 'STITCH_REGISTRY_URL', description: 'Url of the registry' }),
    'authorization-header': command_1.flags.string({ required: false, description: 'Custom authorization header' }),
    timeout: command_1.flags.integer({ required: false, default: 10000, description: 'Request timeout' }),
    verbose: command_1.flags.boolean({ required: false, default: false, description: 'Verbose mode' }),
};
ApplyResources.args = [{ name: 'remoteServerUrl', required: true }];
