// stitch-cli/lib/client/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.refreshRemoteSchema = exports.uploadIntrospectionQueryPolicy = exports.uploadBasePolicy = exports.uploadResourceGroup = void 0;
const tslib_1 = require("tslib");
const graphql_request_1 = require("graphql-request");
const UploadResourceGroupMutation = graphql_request_1.gql `
  mutation UploadResources($resourceGroup: ResourceGroupInput!) {
    result: updateResourceGroup(input: $resourceGroup) {
      success
    }
  }
`;
const ValidateResourceGroupQuery = graphql_request_1.gql `
  query ValidateResources($resourceGroup: ResourceGroupInput!) {
    result: validateResourceGroup(input: $resourceGroup) {
      success
    }
  }
`;
const UploadBasePolicyMutation = graphql_request_1.gql `
  mutation UploadBasePolicyMutation($basePolicy: BasePolicyInput!) {
    result: updateBasePolicy(input: $basePolicy) {
      success
    }
  }
`;
const ValidateBasePolicyQuery = graphql_request_1.gql `
  query ValidateBasePolicyQuery($basePolicy: BasePolicyInput!) {
    result: validateBasePolicy(input: $basePolicy) {
      success
    }
  }
`;
const UploadIntrospectionQueryPolicyMutation = graphql_request_1.gql `
  mutation UploadIntrospectionQueryPolicyMutation($introspectionQueryPolicy: BasePolicyInput!) {
    result: updateIntrospectionQueryPolicy(input: $introspectionQueryPolicy) {
      success
    }
  }
`;
const ValidateIntrospectionQueryPolicyQuery = graphql_request_1.gql `
  query ValidateIntrospectionQueryPolicyQuery($introspectionQueryPolicy: BasePolicyInput!) {
    result: validateIntrospectionQueryPolicy(input: $introspectionQueryPolicy) {
      success
    }
  }
`;
const RefreshRemoteSchemaMutation = graphql_request_1.gql `
  mutation RefreshRemoteSchemaMutation($url: String!) {
    result: refreshRemoteSchema(url: $url) {
      success
    }
  }
`;
function initClient(options, clientOptions) {
    if (typeof options.authorizationHeader !== 'undefined') {
        if (!clientOptions.headers) {
            clientOptions.headers = {};
        }
        clientOptions.headers['Authorization'] = options.authorizationHeader;
    }
    const registryClient = new graphql_request_1.GraphQLClient(options.registryUrl, clientOptions);
    return registryClient;
}
async function uploadResourceGroup(rg, options, clientOptions) {
    const registryClient = initClient(options, clientOptions);
    const query = options.dryRun ? ValidateResourceGroupQuery : UploadResourceGroupMutation;
    return registryClient.request(query, { resourceGroup: rg });
}
exports.uploadResourceGroup = uploadResourceGroup;
async function uploadBasePolicy(basePolicy, options, clientOptions) {
    const registryClient = initClient(options, clientOptions);
    const query = options.dryRun ? ValidateBasePolicyQuery : UploadBasePolicyMutation;
    return registryClient.request(query, { basePolicy });
}
exports.uploadBasePolicy = uploadBasePolicy;
async function uploadIntrospectionQueryPolicy(introspectionQueryPolicy, options, clientOptions) {
    const registryClient = initClient(options, clientOptions);
    const query = options.dryRun ? ValidateIntrospectionQueryPolicyQuery : UploadIntrospectionQueryPolicyMutation;
    return registryClient.request(query, { introspectionQueryPolicy });
}
exports.uploadIntrospectionQueryPolicy = uploadIntrospectionQueryPolicy;
async function refreshRemoteSchema(url, options, clientOptions) {
    const registryClient = initClient(options, clientOptions);
    const query = RefreshRemoteSchemaMutation;
    return registryClient.request(query, { url });
}
exports.refreshRemoteSchema = refreshRemoteSchema;
tslib_1.__exportStar(require("./types"), exports);
